carma_wm_ctrl/src/WMBroadcaster.cpp

```diff
@@ -308,41 +409,55 @@ void WMBroadcaster::addSpeedLimit(std::shared_ptr<Geofence> gf_ptr)
-    }
-    if (gf_ptr->min_speed_limit_->id() != lanelet::InvalId)
+    if (gf_ptr->regulatory_element_->id() != lanelet::InvalId)
     {
-      current_map_->update(current_map_->laneletLayer.get(el.id()), gf_ptr->min_speed_limit_);
-      gf_ptr->update_list_.push_back(std::pair<lanelet::Id, lanelet::RegulatoryElementPtr>(el.id(), gf_ptr->min_speed_limit_));
+      current_map_->update(current_map_->laneletLayer.get(el.id()), gf_ptr->regulatory_element_);
+      gf_ptr->update_list_.push_back(std::pair<lanelet::Id, lanelet::RegulatoryElementPtr>(el.id(), gf_ptr->regulatory_element_));
     }
   }
   
 }
 
-void WMBroadcaster::addBackSpeedLimit(std::shared_ptr<Geofence> gf_ptr)
+void WMBroadcaster::addBackRegulatoryComponent(std::shared_ptr<Geofence> gf_ptr)
 {
   // First loop is to remove the relation between element and regulatory element that this geofence added initially
+  
   for (auto el: gf_ptr->affected_parts_)
   {
     for (auto regem : el.regulatoryElements())
     {
-      if (regem->attribute(lanelet::AttributeName::Subtype).value() == lanelet::DigitalSpeedLimit::RuleName)
+      // extra condition for PassingControlLine
+      if (regem->attribute(lanelet::AttributeName::Subtype).value() == lanelet::PassingControlLine::RuleName && el.isLanelet())
+      {
+        lanelet::PassingControlLinePtr pcl =  std::dynamic_pointer_cast<lanelet::PassingControlLine>(current_map_->regulatoryElementLayer.get(regem->id()));
+        // if this geofence's pcl doesn't match with the lanelets current bound side
+        bool should_process_this_regem = false;
+        for (auto control_line : pcl->controlLine())
+        {
+          if (control_line.id() == el.lanelet()->leftBound2d().id() && gf_ptr->pcl_affects_left_ ||
+          control_line.id() == el.lanelet()->rightBound2d().id() && gf_ptr->pcl_affects_right_)
+          {
+            should_process_this_regem = true;
+            break;
+          }
+        }
+        if (!should_process_this_regem) continue;
+      }
+
+      if (regem->attribute(lanelet::AttributeName::Subtype).value() ==  gf_ptr->regulatory_element_->attribute(lanelet::AttributeName::Subtype).value())
       {
         auto nonconst_regem = current_map_->regulatoryElementLayer.get(regem->id());
         gf_ptr->remove_list_.push_back(std::make_pair(el.id(), nonconst_regem));
         current_map_->remove(current_map_->laneletLayer.get(el.lanelet()->id()), nonconst_regem);
       }
     }
   }
-
+  
   // As this gf received is the first gf that was sent in through addGeofence,
   // we have prev speed limit information inside it to put them back
   for (auto pair : gf_ptr->prev_regems_)
   {
-    if (pair.second->attribute(lanelet::AttributeName::Subtype).value() == lanelet::DigitalSpeedLimit::RuleName)
+    if (pair.second->attribute(lanelet::AttributeName::Subtype).value() ==  gf_ptr->regulatory_element_->attribute(lanelet::AttributeName::Subtype).value())
```

Comment:
In a bunch of your checks you are using == for comparing these attribute values. Please verify this will give you a valid comparison given I think they are stored as strings.